main/src/library/BigInt.flix

```diff
@@ -97,6 +97,234 @@ namespace BigInt {
+    ///
+    /// Returns `None` if the numeric value of `x` is outside the range of Int8
+    /// (i.e. -128 to 127).
+    ///
+    pub def tryToInt8(x: BigInt): Option[Int8] =
+        import java.math.BigInteger.byteValueExact();
+        try {
+            Some(byteValueExact(x)) as & Pure
+        } catch {
+            case _ : ##java.lang.ArithmeticException => None
+        }
+
+    ///
+    /// Convert `x` to an `Option[Int16]`.
+    ///
+    /// Returns `Some(x as Int16)` if the numeric value of `x` can be represented exactly.
+    ///
+    /// Returns `None` if the numeric value of `x` is outside the range of Int16
+    /// (i.e. -32768 to 32767).
+    ///
+    pub def tryToInt16(x: BigInt): Option[Int16] =
+        import java.math.BigInteger.shortValueExact();
+        try {
+            Some(shortValueExact(x)) as & Pure
+        } catch {
+            case _ : ##java.lang.ArithmeticException => None
+        }
+
+    ///
+    /// Convert `x` to an `Option[Int32]`.
+    ///
+    /// Returns `Some(x as Int32)` if the numeric value of `x` can be represented exactly.
+    ///
+    /// Returns `None` if the numeric value of `x` is outside the range of Int32
+    /// (i.e. -2147483648 to 2147483647).
+    ///
+    pub def tryToInt32(x: BigInt): Option[Int32] =
+        import java.math.BigInteger.intValueExact();
+        try {
+            Some(intValueExact(x)) as & Pure
+        } catch {
+            case _ : ##java.lang.ArithmeticException => None
+        }
+
+    ///
+    /// Convert `x` to an `Option[Int64]`.
+    ///
+    /// Returns `Some(x as Int64)` if the numeric value of `x` can be represented exactly.
+    ///
+    /// Returns `None` if the numeric value of `x` is outside the range of Int64
+    /// (i.e. -9223372036854775808 to 9223372036854775807).
+    ///
+    pub def tryToInt64(x: BigInt): Option[Int64] =
+        import java.math.BigInteger.longValueExact();
+        try {
+            Some(longValueExact(x)) as & Pure
+        } catch {
+            case _ : ##java.lang.ArithmeticException => None
+        }
+
+    ///
+    /// Convert `x` to an `Option[Float32]`.
+    ///
+    /// Returns `Some(x as Float32)` if the numeric value of `x` is within the range of Float32.
+    ///
+    /// Warning: even if `x` is in the range of Float32 it may lose precision.
+    ///
+    /// Returns `None` if the numeric value of `x` is outside the range of Float32
+    /// (i.e. 1.4E-45 to 3.4028235E38).
+    ///
+    pub def tryToFloat32(x: BigInt): Option[Float32] =
+        use Option.flatMap;
+        import new java.math.BigDecimal(BigInt) as newBigDecimal;
+        import java.math.BigDecimal.floatValue();
+        let xd = newBigDecimal(x) as & Pure;
+        let d = xd.floatValue() as & Pure;
+        let* f32min = Float32.tryToBigInt(Float32.minValue());
+        let* f32max = Float32.tryToBigInt(Float32.maxValue());
+        if (x < f32min || x > f32max)
```

Comment:
I was tought to write such limits as:

```
if (f32min <= x && x <= f32max) 
  Some(d)
```
because it closely mirrors mathematical syntax of `x < y < z`.